### PR TITLE
API for supporting zebrafish single cell dataset from Kunst et al 2019

### PR DESCRIPTION
**Pull request by vigji**
New API class to download and query by region the single cell tracing dataset from Kunst et al 2019. Currently the entire dataset is downloaded (about 56 MB), a table is generated with soma positions, and query by region will be implemented.

I am happy to hear any suggestion on this organisation and what would provide the smoothest integration with BrainRender!

I also refactored the base Paths class to use pathlib and cleaned up a bit complicated file loading in Neuron class.

### CODE REVIEW

**Comment by FedeClaudi** [comment]
Hey, 

I've only skimmed the code but it looks good. I will have a more detailed look later. 

With regard to brainrender, I can see two options:
1. we do nothing. Morhpapi's `Neuron` class can crate `vedo` actors for the neurons so users could use your api to load data, create the meshes and use `scene.add_actor` to add them to the scene
2. In brainrender we create a class like [ABA](https://github.com/BrancoLab/BrainRender/blob/master/brainrender/atlases/aba.py) which extends the base `Atlas` class endowing it with a `get_neurons` method. This method then handles interaction with your API behind the scenes. 

I would be in favour of 2 for consistency with the phylosophy behind brainrender which aims to limit the number of APIs/softwares that people need to learn to visualize anatomical data. Let me know, what you think!

**Comment by FedeClaudi** [comment]
The code looks great, including the changes to the `Neuron` class, very nicely done!

Not knowing the dataset I have no idea if/how we could get metadata for each neuron (e.g. cell type, brain region of the soma etc) which would let users select which neurons to load. To you have any idea if this kind of metadata can be downloaded somewhere?

----

More generally, in `brainrender` I've introduced to the decorators to check for internet connection for functions that need it and one for checking that the filepath exists for functions like `load_mpin_fixed_neuron`, might be nice to include the same here. Though it doesn't have to be part of this PR.

**Comment by vigji** [comment]
Yes, I would also go for something like 2., it would be definitively better.
It would probably force us to finally change the funny init of the general Base class that happens to be by default an ABA child as well! :) Which is good but also means some dangerous changes to the BrainRender core code.

**Comment by vigji** [comment]
I think this is mergeable now; it supports the creation of a dataframe with the neuron regions onto which an API can be created from `BrainRender`.
If you have nothing against it I would merge!

**Comment by FedeClaudi** [comment]
hey @vigji I'm back from the holidays. Will play around with this tomorrow, but I agree that is seems ready, nice job!

File: morphapi/api/mpin_celldb.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    df.iloc[0, 1] = 1
+    df.iloc[1:, 1] = 2
+
+    # Map points to BrainGlobe orientation:
+    df.iloc[:, 2:-2] = bgspace.map_points_to(TARGET_SPACE, df.iloc[:, 2:-2])
+    df.iloc[0, -2] = NEW_SOMA_SIZE
+    df.to_csv(fixed_file_path, sep=" ", header=None, index=False)
+
+
+class MpinMorphology(Paths):
```
Comment (FedeClaudi) [inline morphapi/api/mpin_celldb.py:59]: can we rename this to MpinMorphologyAPI, for consistency with [other similar classes](https://github.com/brainglobe/morphapi/blob/master/morphapi/api/neuromorphorg.py)

**Comment by FedeClaudi** [comment]
with a fresh install I'm getting an error with `download_daset` (while trying to fix the files):

```
data.zip: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 53.7M/53.7M [00:18<00:00, 2.98MB/s]
Fixing swc files:   0%|                                                                                                                                                                            | 0/1939 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 1654, in create_block_manager_from_blocks
    make_block(values=blocks[0], placement=slice(0, len(axes[0])))
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 3049, in make_block
    return klass(values, ndim=ndim, placement=placement)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 125, in __init__
    f"Wrong number of items passed {len(self.values)}, "
ValueError: Wrong number of items passed 4, placement implies 3

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/federicoclaudi/Documents/Github/morphapi/workspace.py", line 11, in <module>
    api = MpinMorphology()
  File "/Users/federicoclaudi/Documents/Github/morphapi/morphapi/api/mpin_celldb.py", line 69, in __init__
    self.download_dataset()
  File "/Users/federicoclaudi/Documents/Github/morphapi/morphapi/api/mpin_celldb.py", line 127, in download_dataset
    fix_mpin_swgfile(f, self.data_path / f.name)
  File "/Users/federicoclaudi/Documents/Github/morphapi/morphapi/api/mpin_celldb.py", line 54, in fix_mpin_swgfile
    df.iloc[:, 2:-2] = bgspace.map_points_to(TARGET_SPACE, df.iloc[:, 2:-2])
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/bg_space/core.py", line 26, in decorated
    return method(spaceconv_instance, space_description, *args, **kwargs)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/bg_space/core.py", line 344, in map_points_to
    pts_to_transform = np.insert(pts, 3, np.ones(pts.shape[0]), axis=1)
  File "<__array_function__ internals>", line 6, in insert
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/numpy/lib/function_base.py", line 4601, in insert
    return wrap(new)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/generic.py", line 1918, in __array_wrap__
    return self._constructor(result, **d).__finalize__(self)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/frame.py", line 464, in __init__
    mgr = init_ndarray(data, index, columns, dtype=dtype, copy=copy)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/construction.py", line 210, in init_ndarray
    return create_block_manager_from_blocks(block_values, [columns, index])
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 1664, in create_block_manager_from_blocks
    construction_error(tot_items, blocks[0].shape[1:], axes, e)
  File "/Users/federicoclaudi/miniconda3/envs/dev/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 1694, in construction_error
    raise ValueError(f"Shape of passed values is {passed}, indices imply {implied}")
ValueError: Shape of passed values is (362, 4), indices imply (362, 3)
```

File: morphapi/api/mpin_celldb.py (hunk 1 above)
Comment (vigji) [reply morphapi/api/mpin_celldb.py:59]: sure!

**Comment by vigji** [comment]
Funny, I will try this again in a fresh env ad I have no clue what it might be producing that atm

**Comment by FedeClaudi** [comment]
@vigji I've fixed it, was a small thing. 

I've also completed the load neuron function, it works now. 
image


will add an example about it then it's ready for merge

**Comment by FedeClaudi** [comment]
merged and added to release `0.1.1.5`

**Pull request merged**